api/envoy/service/network_ext_proc/v3/network_external_processor.proto

```diff
@@ -0,0 +1,183 @@
+// along with optional metadata.
+message ProcessingRequest {
+  oneof request {
+    option (validate.required) = true;
+
+    // ReadData contains the network data intercepted in the request path (client to server).
+    // This is sent to the external processor when data arrives from the downstream client.
+    ReadData read_data = 1;
+
+    // WriteData contains the network data intercepted in the response path (server to client).
+    // This is sent to the external processor when data arrives from the upstream server.
+    WriteData write_data = 2;
+  }
+
+  // Optional metadata associated with the request.
+  // This can include connection properties, filter configuration, and any other
+  // contextual information that might be useful for processing decisions.
+  //
+  // The metadata is not automatically propagated from request to response.
+  // The external processor must include any needed metadata in its response.
+  config.core.v3.Metadata metadata = 3;
+}
+
+// ProcessingResponse contains the response from the external processing server to Envoy.
+// Each response corresponds to a ProcessingRequest and indicates how the network
+// traffic should be handled.
+message ProcessingResponse {
+  oneof response {
+    option (validate.required) = true;
+
+    // The processed ReadData containing potentially modified data for the request path.
+    // This should be sent in response to a ProcessingRequest with read_data.
+    ReadData read_data = 1;
+
+    // The processed WriteData containing potentially modified data for the response path.
+    // This should be sent in response to a ProcessingRequest with write_data.
+    WriteData write_data = 2;
+  }
+
+  // Status of the processing operation.
+  // This field communicates to Envoy how the data was processed and
+  // what action should be taken with the connection.
+  ProcessingStatus processing_status = 3;
+
+  // Optional metadata associated with the request.
+  // This can include connection properties, filter configuration, and any other
+  // contextual information that might be useful for processing decisions.
+  //
+  // The metadata is not automatically propagated from request to response.
+  // The external processor must include any needed metadata in its response.
+  google.protobuf.Struct dynamic_metadata = 4;
+}
+
+// ReadData represents raw network data intercepted in the request path (client to server).
+message ReadData {
+  // The raw payload data from the request path.
+  bytes data = 1;
+
+  // Indicates whether this is the last data frame in the current direction.
+  // When true, the client has closed the writing side of the connection or
+  // the read operation has completed.
+  //
+  // The external processor should still respond to this message even
+  // if there is no more data expected in this direction.
+  bool end_of_stream = 2;
+}
+
+// WriteData represents raw network data intercepted in the response path (server to client).
+message WriteData {
```

Comment:
Why do we have the same message for ReadData and WriteData?  Do we envision that we will have cases in the future where we may want to add fields only for reads or only for writes?